src/rdf4cpp/rdf/datatypes/registry/util/StaticFlatMap.hpp

```diff
@@ -0,0 +1,76 @@
+#ifndef RDF4CPP_REGISTRY_UTIL_STATICFLATMAP_HPP
+#define RDF4CPP_REGISTRY_UTIL_STATICFLATMAP_HPP
+
+#include <cstddef>
+#include <algorithm>
+#include <array>
+#include <utility>
+#include <cassert>
+
+namespace rdf4cpp::rdf::datatypes::registry::util {
+
+/**
+ * A map with fixed/static (maximum) capacity that can be used in constexpr contexts
+ *
+ * @tparam Key key type
+ * @tparam Value mapped type
+ * @tparam capacity maximum number of mappings that can be contained in the map
+ */
+template<typename Key, typename Value, size_t capacity>
+struct StaticFlatMap {
+    using key_type = Key;
+    using mapped_type = Value;
+    using value_type = std::pair<Key, Value>;
+    using const_reference = value_type const &;
+    using const_pointer = value_type const *;
+    using size_type = size_t;
+    using difference_type = ptrdiff_t;
+    using const_iterator = value_type const *;
+
+private:
+    using storage_t = std::array<value_type, capacity>;
+    storage_t storage;
+    size_t max_init;
+
+public:
+    constexpr StaticFlatMap(std::initializer_list<std::pair<Key, Value>> const inits) noexcept(std::is_nothrow_copy_constructible_v<value_type>)
+        : max_init{inits.size()} {
+
+        assert(inits.size() <= capacity);
+        std::copy(inits.begin(), inits.end(), this->storage.begin());
+    }
+
+    [[nodiscard]] constexpr size_type size() const noexcept {
+        return this->max_init;
+    }
+
+    [[nodiscard]] constexpr size_type max_size() const noexcept {
+        return capacity;
+    }
+
+    constexpr const_iterator begin() const noexcept {
+        return this->storage.begin();
+    }
+
+    constexpr const_iterator end() const noexcept {
+        return this->storage.begin() + this->max_init;
+    }
+
+    constexpr const_iterator find(key_type const &search_key) const noexcept {
+        return std::ranges::find(this->begin(), this->end(), search_key, [](auto const &entry) {
+            return entry.first;
+        });
+    }
+
+    constexpr bool contains(key_type const &search_key) const noexcept {
+        return this->find(search_key) != this->end();
+    }
+
+    constexpr mapped_type const &operator[](key_type const &key) const noexcept {
+        auto it = this->find(key);
+        return it->second;
```

Comment:
out-of-bounds read if key not contained. `std::optional<std::reference_wrapper<mapped_type const &>>` or just use find?

Reply:
I have an idea that is kinda weird and kinda hacky; but it prevents it from compiling if the key is not found and since we are only using it at compiletime anyways it might be an idea(?) I would really like to use `if consteval` here but thats sadly c++23 so I think it's the best I can do to make it not compile

```c++
...
private:
    consteval const_iterator find_ub(key_type const &search_key) const noexcept {
        auto const it = this->find(search_key);

        if (it == this->end()) {
            return this->storage.end();
        }

        return it;
    }

public:
    consteval mapped_type const &operator[](key_type const &key) const noexcept {
        auto const it = this->find_ub(key);
        return it->second;
    }
...

// compiletime error; array index out of bounds
// because undefined behaviour is not allowed in constexpr contexts and has to be reported
inline constexpr LiteralType t = reserved_datatype_ids["hello"]; 
```

Reply:
and for some reason you cannot just static_assert that even if the function is consteval.. so yeah

Reply:
looks good 👍🏻 
maybe add to the class doc that it is only safe to use at compile time.

Reply:
This seems to break support for clang 12 and 13. Please double check that it cannot (easily) worked around.

Reply:
Sorry, didn't notice. Is fixed now